### PR TITLE
ISSUE-3618: Improved performance of arithmetic plus functions

### LINKED ISSUE #3618: Improved performance of arithmetic functions
Discussion: https://github.com/datafuselabs/databend/discussions/3614
Move arithmetic from datavalues to functions.
The type is no longer converted before the calculation, but is converted during the calculation.

Tasks:
- [x] #3620
- [x] #3651

### PR DESCRIPTION
**Pull request by zhyass**
I hereby agree to the terms of the CLA available at: https://databend.rs/policies/cla/

## Summary

Move arithmetic from datavalues to functions.
The type is no longer converted before the calculation, but is converted during the calculation.
Move num_argments and variadic_arguments from Function trait to FunctionFeatures.

## Changelog

- Improvement
- Performance Improvement

## Related Issues

Fixes #3618 

## Test Plan

Unit Tests

Stateless Tests

### CODE REVIEW

File: common/functions/src/scalars/function_factory.rs [hunk 1]
```diff
@@ -37,13 +38,21 @@ use crate::scalars::UdfFunction;
...
 pub struct FunctionFeatures {
     pub is_deterministic: bool,
     pub negative_function_name: Option<String>,
     pub is_bool_func: bool,
     pub is_context_func: bool,
     pub maybe_monotonic: bool,
+    pub num_arguments: usize,
+    // (1, 2) means we only accept [1, 2] arguments
+    // None means it's not variadic function
+    pub variadic_arguments: Option<(usize, usize)>,
```
Comment (flaneur2020) [inline common/functions/src/scalars/function_factory.rs:55]: how about unify `num_arguments` and `variadic_arguments` into a single (usize, usize) in this struct?

calling num_arguments(2) -> set as (2, 2)
calling variadic_arguments(1, 2) -> set as (1, 2)

**Comment by zhyass** [comment]
Performance on Apple M1 16G
Before:
```
mysql> SELECT count() FROM numbers(10000000000) WHERE NOT ignore(number + number);
+-------------+
| count()     |
+-------------+
| 10000000000 |
+-------------+
1 row in set (1.35 sec)
Read 10000000000 rows, 80 GB in 1.345 sec., 7.43 billion rows/sec., 59.47 GB/sec.

mysql> SELECT count() FROM numbers(10000000000) WHERE NOT ignore(toUInt32(number) + number);
+-------------+
| count()     |
+-------------+
| 10000000000 |
+-------------+
1 row in set (2.08 sec)
Read 10000000000 rows, 80 GB in 2.077 sec., 4.81 billion rows/sec., 38.51 GB/sec.

mysql> SELECT count() FROM numbers(10000000000) WHERE NOT ignore(toUInt32(number) + toUInt16(number));
+-------------+
| count()     |
+-------------+
| 10000000000 |
+-------------+
1 row in set (3.22 sec)
Read 10000000000 rows, 80 GB in 3.227 sec., 3.1 billion rows/sec., 24.79 GB/sec.
```

After:
```
mysql> SELECT count() FROM numbers(10000000000) WHERE NOT ignore(number + number);
+-------------+
| count()     |
+-------------+
| 10000000000 |
+-------------+
1 row in set (1.35 sec)
Read 10000000000 rows, 80 GB in 1.350 sec., 7.41 billion rows/sec., 59.26 GB/sec.

mysql> SELECT count() FROM numbers(10000000000) WHERE NOT ignore(toUInt32(number) + number);
+-------------+
| count()     |
+-------------+
| 10000000000 |
+-------------+
1 row in set (1.65 sec)
Read 10000000000 rows, 80 GB in 1.645 sec., 6.08 billion rows/sec., 48.64 GB/sec.

mysql> SELECT count() FROM numbers(10000000000) WHERE NOT ignore(toUInt32(number) + toUInt16(number));
+-------------+
| count()     |
+-------------+
| 10000000000 |
+-------------+
1 row in set (2.35 sec)
Read 10000000000 rows, 80 GB in 2.345 sec., 4.26 billion rows/sec., 34.12 GB/sec.
```

File: common/functions/src/scalars/function_factory.rs (hunk 1 above)
Comment (zhyass) [reply common/functions/src/scalars/function_factory.rs:55]: > how about unify `num_arguments` and `variadic_arguments` into a single (usize, usize) in this struct?
> 
> calling num_arguments(2) -> set as (2, 2) calling variadic_arguments(1, 2) -> set as (1, 2)

A good idea.

Comment (BohuTANG) [reply common/functions/src/scalars/function_factory.rs:55]: > > how about unify `num_arguments` and `variadic_arguments` into a single (usize, usize) in this struct?
> > calling num_arguments(2) -> set as (2, 2) calling variadic_arguments(1, 2) -> set as (1, 2)
> 
> A good idea.

If the changes big in this PR, I suggest we do it in another PR later, create a new issue first to track it.

Comment (zhyass) [reply common/functions/src/scalars/function_factory.rs:55]: > variadic_arguments

Sure, I will create a new issue and PR.

**Review by sundy-li — APPROVED** [review]
/LGTM

Great.

File: common/functions/src/scalars/function_factory.rs (hunk 1 above)
Comment (sundy-li) [reply common/functions/src/scalars/function_factory.rs:55]: I prefer to keep as origin, it's strange to set `(2,2)` for none variadic argument function but most of the functions are none-variadic .

**Comment by BohuTANG** [comment]
/lgtm

A monster PR, but looks great :+1

**Pull request merged**